### PR TITLE
[20351] Fix on_sample_lost notification on best-effort readers for framented samples

### PR DESCRIPTION
**Pull request by MiguelCompany**
## Description

When some fragments are lost on every sample, best-effort readers were not notifying `on_sample_lost` events.

@Mergifyio backport 2.12.x 2.11.x 2.10.x 2.6.x

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added; the added tests pass locally 
- *N/A* Any new/modified methods have been properly documented using Doxygen. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- *N/A* New feature has been added to the `versions.md` file (if applicable).
- *N/A* New feature has been documented/Current behavior is correctly described in the documentation. 
    
- [x] Applicable backports have been included in the description.

## Reviewer Checklist
- [x] The PR has a milestone assigned.
- [x] Check contributor checklist is correct.
- [ ] Check CI results: changes do not issue any warning.
- [ ] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

**Comment by JesusPoderoso** [comment]
@richiprosima please test this

**Comment by JesusPoderoso** [comment]
@richiprosima please test mac 🥲

File: test/blackbox/common/DDSBlackboxTestsListeners.cpp [hunk 1]
```diff
@@ -711,6 +712,42 @@ void sample_lost_test_dw_init(
 
                 return false;
             };
+    testTransport->drop_data_frag_messages_filter_ = [](eprosima::fastrtps::rtps::CDRMessage_t& msg)-> bool
+            {
+                uint32_t old_pos = msg.pos;
+
+                // see RTPS DDS 9.4.5.4 DataFrag Submessage
+                EntityId_t readerID, writerID;
```
Comment (EduPonz) [inline test/blackbox/common/DDSBlackboxTestsListeners.cpp:720]: I'm okey with doing this with `skip-ci`
```suggestion
                EntityId_t readerID;
                EntityId_t writerID;
```

File: test/blackbox/common/DDSBlackboxTestsListeners.cpp [hunk 2]
```diff
@@ -721,19 +758,22 @@ void sample_lost_test_dw_init(
 
 }
 
+template<typename T>
 void sample_lost_test_dr_init(
-        PubSubReader<HelloWorldPubSubType>& reader,
+        PubSubReader<T>& reader,
         std::function<void(const eprosima::fastdds::dds::SampleLostStatus& status)> functor)
 {
+    reader.socket_buffer_size(20 * 1024 * 1024);
```
Comment (EduPonz) [inline test/blackbox/common/DDSBlackboxTestsListeners.cpp:766]: Add a comment on this

**Review by EduPonz — CHANGES_REQUESTED** [review]
I have two minor suggestions that IMO can be done without running CI if that passes

**Comment by EduPonz** [comment]
@Mergifyio backport 2.12.x 2.11.x 2.10.x 2.6.x

**Comment by EduPonz** [comment]
@Mergifyio rebase master

**Comment by EduPonz** [comment]
@richiprosima please test mac test linux test discovery server

File: test/blackbox/common/DDSBlackboxTestsListeners.cpp (hunk 1 above)
Comment (MiguelCompany) [reply test/blackbox/common/DDSBlackboxTestsListeners.cpp]: 6df3ea1

File: test/blackbox/common/DDSBlackboxTestsListeners.cpp (hunk 2 above)
Comment (MiguelCompany) [reply test/blackbox/common/DDSBlackboxTestsListeners.cpp:771]: 6df3ea1

**Comment by JesusPoderoso** [comment]
@richiprosima please test mac test linux test discovery server

**Review by EduPonz — APPROVED** [review]

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Comment by MiguelCompany** [comment]
@mergifyio rebase

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by EduPonz** [comment]
Failures in macOS seem definitely related to the PR. I'm moving this to 2.14

**Comment by MiguelCompany** [comment]
@richiprosima Please test this

**Comment by MiguelCompany** [comment]
@richiprosima Please test mac

**Comment by MiguelCompany** [comment]
@richiprosima please test mac and test windows

**Comment by EduPonz** [comment]
@Mergifyio backport 2.13.x

**Pull request merged**